### ISSUE TITLE
Passing the GLM package over to bincrafters

### ISSUE THREAD

**Issue by dimi309** [bincrafters/community#204]
### Description of Problem, Request, or Question

As discussed with @solvingj and @danimtb, even though I have been maintaining the GLM package in conan center and the packaging changes had been accepted by the glm library developer, I believe that it would be more efficient that the package be maintained as an "external" one within the bincrafters repositories. I have prepared such a version and I would like to request that it be "picked up" by bincrafters:

https://github.com/dimi309/conan-glm

Even though this is a source-only package with no binaries, I am running CI on travis and appveyor to test that its inclusion in projects works well:

https://ci.appveyor.com/project/coding3d/conan-glm
https://travis-ci.org/dimi309/conan-glm

### Package Details (Include if Applicable)
* Package Name/Version: **glm/0.9.8.5**
* Operating System: **Windows, MacOS, Linux**
* Compiler+version: **any**
* Docker image: **all supported by bincrafters**

**Comment by SSE4** [comment]
Hi @dimi309 
might be unrelated, but what do you think about glew packages? like this one https://github.com/nigels-com/glew/tree/master/build/conan

**Comment by dimi309** [comment]
Hi @SSE4 
Yup! I've packaged that one too and it is in the same situation as GLM (only it has not yet been accepted to conan-center) We have also talked with @solvingj about creating an external glew package and adding it to the bincrafters repositories. I will do that next 😄

**Comment by SSE4** [comment]
@dimi309 thank you! it would be nice to have complete OpenGL stuff in conan center

**Comment by dimi309** [comment]
My pleasure! I can guarantee they work pretty well too, because I use them all the time 😉

**Comment by dimi309** [comment]
Thanks for creating the repos! I've added the glm code: https://github.com/bincrafters/conan-glm
Could somebody activate the builds? We can put the badges in the readme afterwards.

**Comment by uilianries** [comment]
@dimi309 I'll take this :+1:

**Comment by uilianries** [comment]
Hey @dimi309 @SSE4 @solvingj !

Please, review https://github.com/bincrafters/conan-glm

https://ci.appveyor.com/project/BinCrafters/conan-glm
https://travis-ci.org/bincrafters/conan-glm
https://bintray.com/bincrafters/public-conan/glm%3Abincrafters

**Comment by uilianries** [comment]
Well, there is the official GLM recipe on Conan center provided by g-truc.

Should we keep a duplicated recipe?

**Comment by dimi309** [comment]
Thanks for looking into this @uilianries! It looks good to me!

In fact, I am the maintainer of the GLM package currently in conan-center. I have just had my packaging code accepted into g-truc's repository and I could publish using his name through my account because I wasn't using CI. But after a discussion we have had with @danimtb I think it is more efficient to be maintaining an external GLM package in bincrafters (see first message on this thread).

So, @danimtb, @lasote, can we replace the package in conan-center with this one?

**Comment by lasote** [comment]
I'm afraid it is not that easy. It can be done but we have to contact bintray people to support us. There are some mechanism to block the complete package deletion in bintray. we need that to be able to link the bincrafters one.

**Comment by dimi309** [comment]
@lasote is this deletion blocking new or only associated with conan-center? I have deleted entire packages on my own repo before and nothing stopped me. In any case, the new "external" package produces a glm@bincrafters signature whereas I was just uploading the old one directly as glm@g-truc, so I think we can survive with a few leftover artifacts...

**Comment by uilianries** [comment]
Bintray could use some warning message when try to install by glm@g-truc, as "This package was moved to glm@bincrafters", or it could be linked directly to  glm@bincrafters.

**Comment by danimtb** [comment]
We have to do some checks as @lasote said. 

The thing here is that it would be nice to have the new package uploaded under the same `user/channel` so users will not notice the change in their recipes but we have some concerns on how the bintray backend will behave. We will try to see if we can unlink the current glm package from conan-center and link the bincrafters one without issue.

I think the message @uilianries is proposing is not possible right now.

**Comment by dimi309** [comment]
By the way, @danimtb & @lasote, while you're looking into this please note that you have my consent to delete my existing glm package from bintray. Or let me know if and when you would like me to do so myself.

**Comment by danimtb** [comment]
I have done some tests with lasote in Bintray and the package an be liked directly without inclusion request, so we aim to replace the old one with this new bincrafters one.

Having a look at the repo it seems the package is ready for the replacement. I have only noticed appveyor badge points to the sqlite jobs 😆 (no rush or stopper, just fix it when you can). 

Final thing would be to have the packages uploaded under the `g-truc` name so as to not brake current users of the old package. How hard would it be to change that in your CI setup?

**Comment by dimi309** [comment]
Hi @danimtb,

I've fixed the appveyor badge.

Uploading under the g-truc name could be done by adding CONAN_USERNAME="g-truc" in the environment variables on CI. I don't have access to do this but maybe @uilianries, @SSE4 or @solvingj can help us there. If that by itself causes any problems maybe we should also add CONAN_LOGIN_USERNAME="bincrafters". I've checked the source code of build_shared.py in bincrafters_package_tools and ConanMultiPackager in conan_package_tools, which is used by the former and it seems to me that what I am suggesting would work.

**Comment by solvingj** [comment]
Yes I think @dimi309 is correct.  I will update the `conan_username` field .

**Comment by dimi309** [comment]
Thanks @solvingj. The build has failed unfortunately. Maybe it also needs CONAN_LOGIN_USERNAME="bincrafters"

**Comment by solvingj** [comment]
The problem is that `bincrafters` is an org not a user.  We use `bincrafters-user` for authentication, and so it occupies the `CONAN_LOGIN_USERNAME` variable.  I guess need a way to override the user in the constructor for this case, which may require a slight tweak to `bincrafters_package_tools`.

**Comment by dimi309** [comment]
I could look into it but I think that someone with more access privileges and more experience managing the package tools could probably find a solution faster.

**Comment by dimi309** [comment]
Thank you!

**Comment by uilianries** [comment]
@SSE4 where did push your changes? I didn't find any commit over https://github.com/bincrafters/conan-glm

**Comment by SSE4** [comment]
@uilianries https://github.com/bincrafters/bincrafters-package-tools/commit/14babc5392702bfd2e915f6174fc77a7464cee0e

**Comment by uilianries** [comment]
@SSE4 your changes weren't pushed into pypi repo, only tags are deployed. I'm going to update now.

**Comment by uilianries** [comment]
new release is done: https://pypi.org/project/bincrafters-package-tools/0.15.0/

**Comment by SSE4** [comment]
unfortunately still didn't help :( I have no idea what exactly goes wrong

**Comment by uilianries** [comment]
I'll take a look later, I'm out of office today

**Comment by uilianries** [comment]
It's possible read the wrong upload address on Appveyor log:
https://ci.appveyor.com/project/BinCrafters/conan-glm#L208

The problem is caused because Bincrafters' package tools assumes the username and repo name as the same:

https://github.com/bincrafters/bincrafters-package-tools/blob/master/bincrafters/build_shared.py#L101

TIL we need to set CONAN_UPLOAD on environment settings. 

![](http://www.callcentermemes.com/wp-content/uploads/2015/12/1449610204609.jpg)

**Comment by SSE4** [comment]
closing, as package was successfully built

**Comment by uilianries** [comment]
@lasote @danimtb The package is done on our side, but the reference on Conan center points to @dimi309 repo. Could you move the reference to our package?

**Comment by danimtb** [comment]
We will change the link to conan center ASAP.

Thanks a lot @uilianries, @SSE4 and @dimi309 for the hard work guys!

**Comment by danimtb** [comment]
Swap done guys! Now the bincrafters package is the one linked to conan-center: https://bintray.com/bincrafters/public-conan/glm%3Ag-truc 🎉 

`conan install glm/0.9.8.5@g-truc/stable` downloads bincrafters recipe instead of the @dimi309 one 😃 

Thanks again for moving this forward!

**Comment by dimi309** [comment]
Wonderful! Thanks @danimtb ! I'm glad we now have a version of the package controlled and supported by the entire bincrafters team!
